### PR TITLE
Graft card state out of HanabiCard

### PR DESCRIPTION
**Pull request by robertos**
Move CardState to types
Add variant pure functions (+ tests)

### CODE REVIEW

File: client/src/game/ui/HanabiCard.ts [hunk 1]
```diff
@@ -462,29 +429,33 @@ export default class HanabiCard extends Konva.Group {
     // If the card is already fully revealed from clues, then additional clues would tell us nothing
     // We don't check for "this.identityDetermined" here because we still need to calculate the
     // effects of clues on cards in other people's hands that we already know the true identity of
-    const wasFullyKnown = this.possibleSuits.length === 1 && this.possibleRanks.length === 1;
+    const wasFullyKnown = this.state.possibleSuits.length === 1
```
Comment (Zamiell) [inline client/src/game/ui/HanabiCard.ts:432]: breaks airbnb style, same as previous thing

File: client/src/game/ui/HanabiCard.ts [hunk 2]
```diff
@@ -804,35 +779,39 @@ export default class HanabiCard extends Konva.Group {
...
-      [this.rank] = this.possibleRanks;
-      globals.learnedCards[this.order].rank = this.rank;
+      [this.state.rank] = this.state.possibleRanks;
+      globals.learnedCards[this.state.order].rank = this.state.rank;
       this.rankPips!.hide();
     }
 
     // Handle if this is the first time that the card is fully revealed to the holder
-    const isFullyKnown = this.possibleSuits.length === 1 && this.possibleRanks.length === 1;
+    const isFullyKnown = this.state.possibleSuits.length === 1
```
Comment (Zamiell) [inline client/src/game/ui/HanabiCard.ts:793]: make paren like other thing

File: client/src/game/ui/HanabiCard.ts (hunk 1 above)
Comment (robertos) [reply client/src/game/ui/HanabiCard.ts:432]: The current linter options doesn't allow
```
const wasFullyKnown = 
  this.state.possibleSuits.length === 1
```

Comment (robertos) [reply client/src/game/ui/HanabiCard.ts:432]: Maybe with a paren. I'll check

**Comment by Zamiell** [comment]
looks good, thank you!!

**Pull request merged**